src/Bech32.mo

```diff
@@ -0,0 +1,296 @@
+      input.size() - 1 - separatorIndex);
+
+    for (i in Iter.range(0, separatorIndex - 1)) {
+      hrpBuffer.add(toLower(inputData[i]));
+    };
+
+    for (i in Iter.range(separatorIndex + 1, inputData.size() - 1)) {
+      let c : Nat8 = inputData[i];
+      let mappedVal : Nat8 = reverseCharset[Nat8.toNat(c)];
+
+      if (mappedVal == 255) {
+        return #err ("Invalid character found: " # toText(c));
+      };
+      valuesBuffer.add(mappedVal);
+    };
+    let values : [Nat8] = valuesBuffer.toArray();
+    let hrp : [Nat8] = hrpBuffer.toArray();
+
+    return switch (verifyChecksum(hrp, values),
+      Text.decodeUtf8(Blob.fromArray(hrp))) {
+        case (#err (msg), _) {
+          #err (msg)
+        };
+        case (#ok (encodingType), ?hrp) {
+          // Remove checksum from data.
+          for (i in Iter.range(0, 5)) {
+            ignore(valuesBuffer.removeLast());
+          };
+          return #ok (encodingType, hrp, valuesBuffer.toArray());
+        };
+        case _ {
+          #err ("Failed to decode HRP.");
+        };
+    };
+  };
+
+  // Expand HRP for checksum computations by grouping together the first 3 bits
+  // of all characters, then a zero then the last 5 bits of all characters.
+  // [a, b] => [a[:3], b[:3]] + [0] + [a[3:], b[3:]].
+  func expandHrp(hrp : [Nat8]) : [Nat8] {
+    let hrpSize = hrp.size();
+    let outputSize = hrpSize * 2 + 1;
+    let output = Array.init<Nat8>(outputSize, 0);
+
+    var i : Nat = 0;
+    for (currHrp in hrp.vals()) {
+      output[i] := currHrp >> 5;
+      output[i + hrpSize + 1] := currHrp & 0x1f;
+      i += 1;
+    };
+
+    return Array.freeze(output);
+  };
+
+  // Constant value associated to the given encoding.
+  func encodingConstant(encoding : Encoding) : Nat32 {
+    return switch(encoding) {
+      case (#BECH32) {
+        1
+      };
+      case (#BECH32M) {
+        0x2bc830a3
+      };
+    };
+  };
+
+  // Compute the checksum values for given hrp and data.
+  func createChecksum(hrp : [Nat8], data : [Nat8], encoding : Encoding) : [Nat8] {
+    let expandedHrp : [Nat8] = expandHrp(hrp);
+
+    // Merge expandedHrp and data arrays and append 6 zeroes to get
+    // [expandedHrp..., data..., 0, 0, 0, 0, 0, 0].
+    let polyModValues : Buffer.Buffer<Nat8> = Buffer.Buffer<Nat8>(
+      expandedHrp.size() + data.size() + 6);
+
+    for (val in expandedHrp.vals()) {
+      polyModValues.add(val);
+    };
+    for (val in data.vals()) {
+      polyModValues.add(val);
+    };
+
+    // Pad with 6 zeros.
+    for (_ in Iter.range(0, 5)) {
+      polyModValues.add(0);
+    };
+
+    let mod : Nat32 = polymod(polyModValues.toArray())
+      ^ encodingConstant(encoding);
+
+    // Convert the 5-bit groups in mod to checksum data.
+    return Array.tabulate<Nat8>(6, func (i) {
+      Nat8.fromIntWrap(Nat32.toNat(
+        (mod >> (5 * (5 - Nat32.fromIntWrap(i)))) & 31))
+    });
+  };
+
+  func verifyChecksum(hrp : [Nat8], values : [Nat8])
```

Comment:
Missing documentation.

Reply:
I thought private funcs are fine without docs at least if the name is self explanatory, I'll add it.

Reply:
Yeah, you are right. Maybe it's just me but I find it helpful if every function at least comes with a short line that says what the function does.
If it's very obvious what it does and it's a private function, I can live without a documentation string.

Reply:
No prob, I added one.